### PR TITLE
controllers: ION Discover DJ Pro

### PR DESCRIPTION
**Pull request by meltedpianoman**
As requested by https://github.com/mixxxdj/mixxx/pull/2893

I add the ION Discover DJ Pro hardware to the controller hardware section in the Mixxx manual.

### CODE REVIEW

**Comment by meltedpianoman** [comment]
@Holzhaus : Can you please have a look?
Thanks!

**Comment by Holzhaus** [comment]
Thanks. Here are some vector images of the Controller: [ion_discover_dj_pro.zip](https://github.com/mixxxdj/manual/files/5461429/ion_discover_dj_pro.zip)

Can you add numbered labels with Inkscape and then document the controls like on the [Pioneer DDJ-200](https://manual.mixxx.org/2.3/en/hardware/controllers/pioneer_ddj_200.html) page? The font for the labels is "Open Sans".

**Comment by meltedpianoman** [comment]
Hi @Holzhaus , I updated the manual as you requested similar to the Pioneer DDJ-200 with a separate Deck & Mixer section.

Can you please have a look and give me your comments on this?

Thanks,
Ivo

**Comment by Holzhaus** [comment]
Can you please open the SVG files in Inkscape and save them as "Optimized SVG" to reduce line count and file size?

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+3    :hwlabel:`A CUE` / :hwlabel:`B CUE` buttons              Sends the selected channel's audio to the headphone output
+===  ======================================================   ============================================================================================
+
+Known Issues
+------------
+
+Controls not yet working (or tested):
+
+- Library browsing
+- Cue control
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:92]: What do mean by that? The controller does not have hotcue pads (judging from photos I found on Google)

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+2    Cross fader                                              Fades between left and right deck.
+3    :hwlabel:`A CUE` / :hwlabel:`B CUE` buttons              Sends the selected channel's audio to the headphone output
+===  ======================================================   ============================================================================================
+
+Known Issues
+------------
+
+Controls not yet working (or tested):
+
+- Library browsing
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:91]: I think mapping the browse knob and the deck load buttons is a critical feature and shouldn't be hard to map). Could you take care of that?

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+- Cue control
+- Autoloop
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:93]: This should be easy to map. Check the `beatloop_X_activate` controls.

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 1 above)
Comment (Holzhaus) [reply source/hardware/controllers/ion_discover_dj_pro.rst:92]: Or are you referring to PFL?

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+===  ======================================================   ============================================================================================
+No.  Control                                                  Function
+===  ======================================================   ============================================================================================
+1    :hwlabel:`JOG WHEEL`                                     Scratches the track
+2    :hwlabel:`PLAY/PAUSE` button                             Plays/pauses a track in the respective Deck
+3    :hwlabel:`SYNC` button                                   Sync the tempo to other playing track
+4    :hwlabel:`SCRATCH` button                                Toggle the scratch mode
+5    :hwlabel:`TEMPO` slider                                  Adjust the track playing speed
+6    :hwlabel:`PITCH BEND` buttons -/+                        Lower (-) or Raise (+) the pitch temporarily
+7    :hwlabel:`REVERSE` button                                Hold for reverse play
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:61]: The button is labeled "Reverse Play" I think.

**Review by Holzhaus — COMMENTED** [review]
Thank you. 

What about the STUTTER buttons, the ALL MUTE buttons and the CUE VOLUME/MASTER VOLUME knobs?

These are currently not documented.

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 2 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:91]: Welll... indeed the browse button + deck load buttons seems to work. So I can document that. I'll update the documentation on that one.

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 4 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:61]: Yes, you are right, I'll update this text.

**Comment by meltedpianoman** [comment]
> 
> 
> Thank you.
> 
> What about the STUTTER buttons, the ALL MUTE buttons and the CUE VOLUME/MASTER VOLUME knobs?
> 
> These are currently not documented.

Hi @Holzhaus, 

The CUE volume & Master volume knobs indeed control the audio of the two internal sound output channels, so I indeed will document that.

I understand the controller mapping is far from perfect now. What I mainly did was taking the intial effort from [joachim](https://mixxx.discourse.group/t/ion-discover-dj-pro-mapping-script/12120) and updated that one to the new XML scheme, inverted some sliders and fixed a filepath that did not work on a case sensitive file system like on an Linux machine.

I'm personally not a DJ, but I'm trying to get this controller supported in Mixxx so my son can use it (also with future versions of the Mixxx software). I was hoping that even though it is not complete and perfect, it at least is a first start to get this controller officially supported by Mixxx. Maybe other users that are missing certain functionality will make this controller map a little bit better in the future... My guess is that still some users are glad to have this controller mapping with at least some basic functionality, instead of no support for this controller at all.

The STUTTER & ALL MUTE buttons are currently not mapped at all. I propose that I list that in the "not yet working" section, together with the AUTOLOOP buttons. 

Is it acceptable by you to accept the current state of this controller mapping, even with its shortcomings? If yes: then I'll update the manual with your remarks and upload it for another review. If not: then I'm afraid that I'm going to give up on this Pull Request because I find it difficult to sufficient find time to contribute further on this. Don't get me wrong: I understand that you have to guard the quality of the mixxx software. So if the status of this controller is not acceptable, I'll can accept that. 

Regards,
Ivo

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 1 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:92]: On 2nd thought: I compared the function of the Cue button on the controller with the Cue button in the Mixxx UI. And they seem to work similar (holding the Cue button plays the track from a fixed location... right?). So I guess the Cue button is actually working as it should be. It just didn't make sense to me... I'll update the documentation for that button as well.

Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:92]: Done

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 2 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:91]: Done

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 3 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:93]: See my remarks below... I really would like to finish this controller mapping and leave some features that are not yet working for the community to work on further... I hope you can live with that.

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 4 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:61]: Done

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+===  ======================================================   ============================================================================================
+
+Known Issues
+------------
+
+Controls not yet working (or untested):
+
+- STUTTER
+- ALL MUTE
+- AUTOLOOP
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:98]:
```suggestion
Controls not yet working:

- :hwlabel:`STUTTER` button
- :hwlabel:`ALL MUTE` button
- :hwlabel:`AUTOLOOP` buttons
```

Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst]: Done

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+.. figure:: ../../_static/controllers/ion_discover_dj_pro.jpg
+   :align: center
+   :width: 100%
+   :figwidth: 100%
+   :alt: ION Discover DJ Pro
+   :figclass: pretty-figures
+
+   ION Discover DJ Pro
+
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:12]: We don't have permission to use this image and it's also low-res and doesn't add information. Please remove (and also the image file).

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+This controller is USB MIDI class compliant, so it is plug-and-play on all platforms without having to install
+any special drivers. The manufacturer’s page only mentions Windows and Mac but it’s confirmed working on Linux
+as well.
+
+Controller Mapping
+------------------
+
+.. note:: Only controls are documented that are known to be working.
+
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:39]: This should be removed, this is redundant with the "Known Issues" section.

```suggestion
```

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Sorry for taking so long to review. Already looks good, just two removals, then we can merge IMHO. Thanks you.

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 6 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:12]: Done

File: source/hardware/controllers/ion_discover_dj_pro.rst (hunk 7 above)
Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:39]: Done

**Review by Holzhaus — APPROVED** [review]
LGTM, thank you!

File: source/hardware/controllers/ion_discover_dj_pro.rst [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+   :width: 100%
+   :figwidth: 100%
+   :alt: ION Discover DJ Pro (schematic view)
+   :figclass: pretty-figures
+
+   ION Discover DJ Pro (schematic view)
+
+-  `Mixxx User Forum <https://mixxx.discourse.group/t/ion-discover-dj-pro-mapping-script/12120>`__
+
+.. versionadded:: 2.2.5
```
Comment (Holzhaus) [inline source/hardware/controllers/ion_discover_dj_pro.rst:17]: Oops, there will be no 2.2.5, I'll change this to 2.3.

Comment (meltedpianoman) [reply source/hardware/controllers/ion_discover_dj_pro.rst:17]: Ok, thank @Holzhaus 

But.. what about the GitHub wiki?
Is that no longer maintained?
- I changed the version number also in the table on https://github.com/mixxxdj/mixxx/wiki/Hardware-Compatibility
- What about the https://github.com/mixxxdj/mixxx/wiki/Ion-Discover-DJ-Pro page?

**Pull request merged**